### PR TITLE
[RPC] Add id's to build errors

### LINKED ISSUE #4635: Unique ID's for Errors reported by RPC
This is something that I discussed earlier with @cwong-ocaml and we agreed to address it eventually.
It would be helpful if build errors reported came with a unique ID. This would be useful for the clients to maintain a table keyed by such a ID to efficiently handle incoming add/remove events.

One simple idea would be for dune to generate a GUID for every error. Slightly more efficient would be to generate a GUID for a session and then append an incrementing integer to every error.

cc @jeremiedimino

### PR DESCRIPTION
**Pull request by rgrinberg**
Fix #4635

### CODE REVIEW

**Review by cwong-ocaml — APPROVED** [review]
Looks fine to me.

**Review by jeremiedimino — APPROVED** [review]
me too

**Comment by aalekseyev** [comment]
Sorry, should have looked at this on Friday, when working on #4634.
Even though there's two code paths the exceptions flow through now, it seems I was able to merge them nicely (id is generated in report_early_exn, but not in reraise_exn).

There's presumably further improvement we'll want to make: if the build reports the same errors, then they should get the same ids so we can avoid clearing and re-creating them. This could be arranged by moving the id into the Memo.Error.t type.
Then we could use those ids to deduplicate errors instead of relying on polymorphic equality, too.
Anyway, the code seems fine without that.

File: otherlibs/dune-rpc/private/dune_rpc_private.ml [hunk 1]
```diff
@@ -247,25 +260,30 @@ module Diagnostic = struct
...
+
   let sexp_severity =
     let open Conv in
     enum [ ("error", Error); ("warning", Warning) ]
 
   let sexp =
     let open Conv in
-    let from { targets; message; loc; severity; promotion; directory } =
-      (targets, message, loc, severity, promotion, directory)
+    let from { targets; message; loc; severity; promotion; directory; id } =
```
Comment (aalekseyev) [inline otherlibs/dune-rpc/private/dune_rpc_private.ml:271]: Not gonna pretend I understand what's going on, and there's no need to delay the PR on this, but I'm surprised this patch can avoid bumping rpc versions or an equivalent thing. Does the sexp format, somehow, not need to be stable?

Are we making an assumption that the reader and the writer of the sexp will be the same dune version?

Even if we assume that, shouldn't we have a check so we can complain if they are not the same version, after all?

Comment (rgrinberg) [reply otherlibs/dune-rpc/private/dune_rpc_private.ml:271]: We do have a mechanism for versioning fields, but we haven't released any stable version of RPC yet, so all the current changes do not need version checks or compatibility layers.

**Comment by rgrinberg** [comment]
@aalekseyev thanks for resolving the merge conflicts

> This could be arranged by moving the id into the Memo.Error.t type.

Yes, that sounds like a good idea.

**Pull request merged**